### PR TITLE
Properly handle twitch api encoding and charset

### PR DESCRIPTION
**Pull request by falstro**
Content-Encoding declares stuff like "gzip" or "deflate", and it appears twitch now uses gzip even if there's no Accept-Encoding header causing the current code to fail with an unknown charset "gzip". I've added the gzip decoding, and also proper parsing of the payload charset in case it would ever be anything non-ascii.

### CODE REVIEW

**Comment by IllusionaryOne** [comment]
As indicated in the template - show test results.  In this case, also show the initial errors, as I do not see any such errors in my production bot and I imagine if the Twitch API wasn't working at all we would have tickets related to it.

You will need to show every Twitch API call failing prior to the change and passing after the change to be considered.

Also, not sure where you are seeing GZip, can you please show that in some requests?  I just ran a couple and see the following from Twitch which does not include the gzip charset.

```
> GET /kraken/users?login=illusionaryone HTTP/1.1
> Host: api.twitch.tv
> User-Agent: curl/7.43.0
> Accept: application/vnd.twitchtv.v5+json
> Client-ID: *myclient-id*
>
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=utf-8
< Server: nginx
< Access-Control-Allow-Origin: *
< Cache-Control: no-cache, no-store, must-revalidate, private
< Expires: 0
< Pragma: no-cache
< Twitch-Trace-Id: 44f73da5e5e065d6839e0cfff12278e5
< X-Ctxlog-Logid: 1-5aa72d5c-9ea2878280021790235e25b0
< Front-End-Https: on
< Timing-Allow-Origin: https://www.twitch.tv
< Date: Tue, 13 Mar 2018 01:46:04 GMT
< Content-Length: 434
< Connection: keep-alive
< Set-Cookie: AWSALB=osl2Pkgd138N9YPh0IcbovKDnQgAyRU/EBxgDY+b6DX8cKOLR8jSMJS1ZB3Aw+MglaIgecMunu8gA+t27dZio12BbXGpNKwc3vE9xlAjYS7YICWpdme3oQFMGNC3; Expires=Tue, 20 Mar 2018 01:46:04 GMT; Path=/
```

**Comment by IllusionaryOne** [comment]
Running with some debug statements added; if you can provide something similar that would be helpful.  Thanks!

```
[03-12-2018 @ 20:04:36.987 MDT] >> Twitch API Info: Code 200 | Encoding null | Type application/json; charset=utf-8
[03-12-2018 @ 20:04:36.987 MDT] >> {"_total":1,"users":[{"display_name":"NotIllusionaryOne","_id":"90595045","name":"notillusionaryone","type":"user","bio":null,"created_at":"2015-05-09T00:08:04.864402Z","updated_at":"2018-02-20T23:31:11.684105Z","logo":"https://static-cdn.jtvnw.net/user-default-pictures/49988c7b-57bc-4dee-bd4f-6df4ad215d3a-profile_image-300x300.jpg"}]}
[03-12-2018 @ 20:04:37.055 MDT] >> Twitch API Info: Code 200 | Encoding null | Type application/json; charset=utf-8
[03-12-2018 @ 20:04:37.056 MDT] >> {"_total":1,"users":[{"display_name":"NotIllusionaryOne","_id":"90595045","name":"notillusionaryone","type":"user","bio":null,"created_at":"2015-05-09T00:08:04.864402Z","updated_at":"2018-02-20T23:31:11.684105Z","logo":"https://static-cdn.jtvnw.net/user-default-pictures/49988c7b-57bc-4dee-bd4f-6df4ad215d3a-profile_image-300x300.jpg"}]}
[03-12-2018 @ 20:04:37.176 MDT] >> Twitch API Info: Code 200 | Encoding null | Type application/json; charset=utf-8
[03-12-2018 @ 20:04:37.176 MDT] >> {"_total":1,"users":[{"display_name":"IllusionaryBot","_id":"106842370","name":"illusionarybot","type":"user","bio":"I am a bot. I do interesting things in chat.","created_at":"2015-11-11T19:44:48.318767Z","updated_at":"2017-09-08T04:01:36.376694Z","logo":"https://static-cdn.jtvnw.net/jtv_user_pictures/illusionarybot-profile_image-a940753bc17fed05-300x300.png"}]}
[03-12-2018 @ 20:04:37.415 MDT] >> Twitch API Info: Code 200 | Encoding null | Type application/json
[03-12-2018 @ 20:04:37.415 MDT] >> {
  "_links": {},
  "chatter_count": 1,
  "chatters": {
    "moderators": [
      "scaniatv"
    ],
    "staff": [],
    "admins": [],
    "global_mods": [],
    "viewers": []
  }
}

[03-12-2018 @ 20:04:37.541 MDT] >> Twitch API Info: Code 200 | Encoding null | Type application/json; charset=utf-8
[03-12-2018 @ 20:04:37.542 MDT] >> {"stream":null}
[03-12-2018 @ 20:04:37.589 MDT] Connected to Twitch Moderation Data Feed
[03-12-2018 @ 20:04:37.645 MDT] >> Twitch API Info: Code 200 | Encoding null | Type application/json; charset=utf-8
[03-12-2018 @ 20:04:37.645 MDT] >> {"stream":null}
[03-12-2018 @ 20:04:38.931 MDT] >> Twitch API Info: Code 200 | Encoding null | Type application/json; charset=utf-8
[03-12-2018 @ 20:04:38.931 MDT] >> {"mature":false,"status":"Test","broadcaster_language":"en","display_name":"NotIllusionaryOne","game":"Forza Horizon 2","language":"en","_id":"90595045","name":"notillusionaryone","created_at":"2015-05-09T00:08:04Z","updated_at":"2018-02-20T23:31:11Z","partner":false,"logo":"https://static-cdn.jtvnw.net/user-default-pictures/49988c7b-57bc-4dee-bd4f-6df4ad215d3a-profile_image-300x300.jpg","video_banner":null,"profile_banner":null,"profile_banner_background_color":null,"url":"https://www.twitch.tv/notillusionaryone","views":375,"followers":5,"broadcaster_type":"","description":""}
[03-12-2018 @ 20:04:40.208 MDT] >> Twitch API Info: Code 200 | Encoding null | Type application/json; charset=utf-8
[03-12-2018 @ 20:04:40.208 MDT] >> {"communities":[]}
```

**Comment by falstro** [comment]
Very well, do you agree that the charset is passed through the Content-Type header, and not trhough the Content-Encoding header (visible in your traces too)? If you look in the [HTTP spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.11) it also lists encoding as gzip (and deflate, and all that good stuff) - so by inspection alone the use of the Content-Encoding as a character encoding is buggy.

It appears wrong for twitch to send gzip encoded data if there's no accept-encoding gzip header - and indeed for the vast majority it doesn't - but there might be something in the newer http specs that allow that, don't know for sure (my http spec knowledge is a bit old). The one issue I had was with the subscriber count replacement in the custom commands (which goes through GetChannelSubscribers - name off the top of my head, might be different), and *that* returned gzip encoding for me. No idea why it did that in this particular case - I even tried explicitly setting "Accept-Encoding: identity" to try and force uncompressed but it didn't help.

I'll see if I can get you the traces you want.

**Comment by falstro** [comment]
Side note, it might be worthwhile to set Accept-Encoding: gzip to lower the bandwidth use, json compresses very well. But it's unrelated to this particular issue obviously.

**Comment by falstro** [comment]
I was able to reproduce it only if the oauth token was invalid- it appears to work with a valid oauth token. It didn't occur to me to test it again with a valid oauth token, because it was only with the fix that I realized it was invalid. My bad. Anyway here's the output (using the fix)

Twitch API Info: Code: 401 | Encoding: gzip | Type: application/json; charset=utf-8
{"error":"Unauthorized","status":401,"message":"Token invalid or missing required scope"}

Without the fix it exits with an exception due to the unsupported character encoding ("gzip") in IOUtils.toString and the error from the API is lost.

The minimal test code:
```
import com.gmt2001.TwitchAPIv5;
import org.json.JSONObject;

public class TestTwitch {
  public static void main(String args[]) {
    TwitchAPIv5 twapi = TwitchAPIv5.instance();
    twapi.SetClientID("<client id from code>");
    twapi.SetOAuth("<broken oauth>");
    twapi.GetChannelSubscriptions("<channel>", 100, 0, true); // called the same way in the code
  }
}
```

**Comment by BrandenB** [comment]
I've seen this error come up a few times in my logs as well, even with a proper OAuth token. It might be some weird logic on Twitch's end.

I will approve once you update your last if statement with braces `{ }` to follow our code guidelines.

**Comment by falstro** [comment]
Upps, my bad, missed that one.

**Comment by BrandenB** [comment]
Looks good, I'll just let @IllusionaryOne look over your new comments and let him merge if he's okay with this change.

Thanks!

File: source/com/gmt2001/TwitchAPIv5.java [hunk 1]
```diff
@@ -132,7 +133,27 @@ private JSONObject GetData(request_type type, String url, String post, String oa
...
                 content = "{}";
             } else {
-                content = IOUtils.toString(i, c.getContentEncoding());
+                // default to UTF-8, it'll probably be the best bet if there's
+                // no charset specified.
+                String charset = "utf-8";
+                String ct = c.getContentType();
+                if (ct != null) {
+                    String[] cts = ct.split(" *; *");
+                    if (cts.length > 1) {
```
[the remark this answers is not in the document]
Comment (BrandenB) [reply source/com/gmt2001/TwitchAPIv5.java]: You can remove it if you'd like. :]

**Pull request merged**